docs: changes to store and testing-store live examples and explanations

sassypianist98 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x]  The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

Changed the live examples in the store and testing-store documentation to align with example-app, which provides greater continuity in learning to users. The respective documentation has also been altered to align with the new examples.

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information

brandonroberts commented:
Thank you @sassypianist98! I will take a look this week.

alex-okrushko commented:
@sassypianist98 Thanks for the PR!
Could you please cd into `./projects/ngrx.io` and run `yarn prettier` there? There would be a number of files that could not be adjusted (because they are read-only), but it should adjust all the files that you are adding into the example. 🙂

Thank you so much for the PR, btw! I'll be taking a look as well.

sassypianist98 commented:
> @sassypianist98 Thanks for the PR!
> Could you please cd into `./projects/ngrx.io` and run `yarn prettier` there? There would be a number of files that could not be adjusted (because they are read-only), but it should adjust all the files that you are adding into the example. 🙂
> 
> Thank you so much for the PR, btw! I'll be taking a look as well.

Hi, thank you so much to both of you for taking a look! Just ran `yarn prettier`--please let me know if you need anything else.

alex-okrushko commented:
1. Well, we had to address it sooner or later - the folder structure 😆 @timdeschryver @brandonroberts 
I strongly believe that we should keep actions to were they are dispatched from (literally in the file next to the component file or effect file). We had this discussion in [Angular Discord - ngrx channel](https://discordapp.com/channels/748677963142135818/748804900565024768/750175656205090847)

2. I file the "whole file" live examples to be a bit too much for the "testing" section. I think it's better to break them apart into smaller chucks, and the docs support that with `regions`. Here are the instructions: https://v9.angular.io/guide/docs-style-guide#displaying-a-code-fragment

sassypianist98 commented:
> 1. Well, we had to address it sooner or later - the folder structure 😆 @timdeschryver @brandonroberts
>    I strongly believe that we should keep actions to were they are dispatched from (literally in the file next to the component file or effect file). We had this discussion in [Angular Discord - ngrx channel](https://discordapp.com/channels/748677963142135818/748804900565024768/750175656205090847)
> 2. I find the "whole file" live examples to be a bit too much for the "testing" section. I think it's better to break them apart into smaller chucks, and the docs support that with `regions`. Here are the instructions: https://v9.angular.io/guide/docs-style-guide#displaying-a-code-fragment

Hi Alex, 

Thank you for taking a look. I'll work on 1. reorganizing the folder structures and 2. reworking the code fragments into more manageable parts. Please let me know if there's anything else you would like me to change. Thank you so much!

brandonroberts commented:
I'm on board with cleaning up the examples but still don't agree with the file organization for the 95% of us outside of Google😉

sassypianist98 commented:
Hi all! I made some changes to the testing store docs code fragments--please let me know if/when you'd like me to make the folder structure changes.

alex-okrushko commented on projects/ngrx.io/content/guide/store/index.md:
```diff
@@ -27,93 +27,89 @@ The following diagram represents the overall general flow of application state i
 - Registered the global state container that is available throughout your application.
-- Injected the `Store` service to dispatch actions and select the current state of the counter.
+- Defined the state, as well as selectors that retrieve specific parts of the state.
+- Created two  distinct components, as well as a service that fetches from the Google Books API. 
```
```suggestion
- Created two distinct components, as well as a service that fetches from the Google Books API. 
```

alex-okrushko commented on projects/ngrx.io/content/guide/store/testing.md:
```diff
@@ -83,44 +83,25 @@ Try the <live-example name="testing-store"></live-example>.
-  });
-});
-</code-example>
+The following example tests the `booksReducer` from the [getting started tutorial](guide/store#tutorial). We run the first test to check that the initial state is returned when the reducer is called with an unknown action. The second test checks that this specific reducer updates the state when the corresponding action is passed which, in this case, is `retrievedBookList`.
```
```suggestion
The following example tests the `booksReducer` from the [getting started tutorial](guide/store#tutorial). In the first test we check that the state returns the same reference when the reducer is not supposed to handle the action (unknown action). The second test checks that `retrievedBookList` action updates the state and returns the new instance of it.
```

alex-okrushko reviewed (changes requested):
Thanks again for working on this 👍 Really appreciate it! 😀

We'll leave the folder structure discussion for some better times 🙂
 
I'm far from the best reviews for the grammar, but this reads better to me.

Also, I think it's best to introduce immutable arrays here. `Entities`, of course, are the better alternative, but at least `ReadonlyArray` would prevent from frequent mistakes early.

bbonczek commented on projects/ngrx.io/content/examples/store/src/app/state/allBooks.selectors.ts:
```diff
@@ -0,0 +1,18 @@
+  (books: Array<Book>) => books
+);
+
+export const selectCollectionIds = (state: AppState) => state.collection;
```
Hi @sassypianist98, I come from #2706 and @timdeschryver told me that in this PR I may find some answers :)

I can see that both `selectBooks` and `selectCollectionIds` select a single piece of state, and yet one of them use createSelector, and the other one does not. Why is the reason to do so? My current understanding of createSelectors function is that you use is whenever you want to use memoization (which for me seems to be almost always), or when you want to have a selector for multiple pieces of state.

I can see 2 options for this file (they are mutually exclusive 😄):

* Option A

 `export const selectCollectionIds = (state: AppState) => state.collection;`

becomes

```
export const selectCollectionIds = createSelector(
  (state: AppState) => state.collection,
  (collectionIds: Array<string>) => collectionIds
);
```
Having that, `selectCollectionIds` benefits by receiving support for memoization. It can be exposed to and used by effects and components now.


* Option B

 `export const selectCollectionIds = (state: AppState) => state.collection;`

becomes

`const selectCollectionIds = (state: AppState) => state.collection;`

Having that, `selectCollectionIds` still does not utilize memoization, but it now considered 'private' selector - it is hidden from components and effects.

brandonroberts replied:
For clarity, the top-level feature key should always use `createFeatureSelector`. 

```ts
export const selectCollectionState = createFeatureSelector<AppState, ReadonlyArray<string>('collection');
```

There is only a single piece of state here, but in normal practice, it would be an object that has multiple properties on it, each one building off the feature selector. We should keep the example straightforward while allowing for more guidance on usage going forward.

brandonroberts commented on projects/ngrx.io/content/examples/store/src/app/state/allBooks.selectors.ts:
```diff
@@ -0,0 +1,18 @@
+import { AppState } from './state';
+import { Book } from '../book-list/books.service';
+
+export const selectBooks = createSelector(
+  (state: AppState) => state.books,
+  (books: Array<Book>) => books
+);
```
Use a feature selector here instead

```suggestion
export const selectBooksState = createFeatureSelector<AppState, ReadonlyArray<Book>('books');
```

brandonroberts commented on projects/ngrx.io/content/examples/testing-store/src/app/state/allBooks.selectors.ts:
```diff
@@ -0,0 +1,18 @@
+import { AppState } from './state';
+import { Book } from '../book-list/books.service';
+
+export const selectBooks = createSelector(
```
Use a feature selector here as mentioned above

brandonroberts commented on projects/ngrx.io/content/examples/testing-store/src/app/state/collection.reducer.spec.ts:
```diff
@@ -0,0 +1,49 @@
+
+describe('CollectionReducer', () => {
+  describe('unknown action', () => {
+    it('should return the default state', () => {
```
```suggestion
    it('should return the previous state', () => {
```

brandonroberts commented on projects/ngrx.io/content/guide/store/testing.md:
```diff
@@ -67,11 +67,11 @@ describe('Auth Guard', () => {
 Usage:
 
-<code-example header="user-greeting.component.ts" path="testing-store/src/app/user-greeting.component.ts"></code-example>
+<code-example header="src/app/state/allBooks.selectors.ts" path="testing-store/src/app/state/allBooks.selectors.ts"></code-example>
```
```suggestion
<code-example header="src/app/state/books.selectors.ts" path="testing-store/src/app/state/books.selectors.ts"></code-example>
```

brandonroberts commented on projects/ngrx.io/content/examples/store/src/app/app.component.html:
```diff
@@ -1,5 +1,8 @@
-<!-- #enddocregion counter -->
\ No newline at end of file
+<h2>Books</h2>
+<app-book-list class="book-list" [books]="books$ | async" (add)="add($event)"></app-book-list>
```
```suggestion
<app-book-list class="book-list" [books]="books$ | async" (add)="onAdd($event)"></app-book-list>
```

brandonroberts commented on projects/ngrx.io/content/examples/store/src/app/app.component.html:
```diff
@@ -1,5 +1,8 @@
+<app-book-list class="book-list" [books]="books$ | async" (add)="add($event)"></app-book-list>
+
+<h2>My Collection</h2>
+<app-book-collection class="book-collection" [books]="bookCollection$ | async" (remove)="remove($event)">
```
```suggestion
<app-book-collection class="book-collection" [books]="bookCollection$ | async" (remove)="onRemove($event)">
```

brandonroberts reviewed (changes requested):
Thanks for working on this!

brandonroberts commented:
@sassypianist98 I really like what you've done here. One thing I'm concerned about is the tutorial being too long. I think we should consider scoping down only changing to the testing section, and leaving the tutorial example simple and focused. What do you think?

cc: @alex-okrushko @timdeschryver

timdeschryver commented:
From my experience I see devs skipping the tutorial (or quickly browse through it) and then take a look at the example app. The reasoning is because they find the tutorial to simple and find the example app more "real world"-y. But then again, the example app just shows the code and not the how's and the why's.

I do get your point tho, for new-comers it might be overwhelming.
Would it be too much to have both examples?

- keep the tutorial example simple and focused
- create a "hands-on" example to build a book store (we could use sections from this example app across the docs)

sassypianist98 commented:
Hi @brandonroberts and @timdeschryver. Thank you so much for your feedback and for waiting! To me, the current counter tutorial helped me understand using the store at a more basic level, but I did look more to the example app to see how to integrate these concepts in a more “real world” example. I’ve found similarly more complex tutorials like the Angular Tour of Heroes more helpful when contextualizing new ideas, so I think it would be beneficial to have a more in-depth tutorial. That's what I think, but would you prefer having both examples instead? Or just the modifications to the testing store section?

brandonroberts commented:
I think having a quick tutorial along with a more extensive hands-on exercise would be good.

sassypianist98 commented:
> I think having a quick tutorial along with a more extensive hands-on exercise would be good.

Should I rework the store page to have both included?

brandonroberts commented:
@sassypianist98 I think we should have a separate page for the walkthrough, so copy the changes you made to a new markdown file for the walkthrough, and undo the changes to the current tutorial. Then add a link to the navigation.json to the walkthrough page.

alex-okrushko commented:
Many thanks @sassypianist98 - we really appreciate our effort and continuous engagement through this review 🙂

sassypianist98 commented:
Hi, @brandonroberts and @alex-okrushko, I updated the branch so that the walkthrough has its own page. I also implemented the changes that you both requested. Is there anything else you would like me to add/

brandonroberts commented:
@sassypianist98 thank you again for all the work done here! I think it looks much better now. There are some minor cleanup things to be done, but those don't need to hold this from getting merged.

sassypianist98 commented:
@brandonroberts @alex-okrushko Thank you both so much for all your feedback and help towards this PR 🙏  🎉
